src/Features/LanguageServer/Protocol/Handler/Hover/HoverHandler.cs

```diff
@@ -51,44 +52,70 @@ public HoverHandler()
                 return null;
             }
 
-            var hover = await GetHoverAsync(info, document, context.ClientCapabilities, cancellationToken).ConfigureAwait(false);
-            return hover;
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            return await GetHoverAsync(info, text, document.Project.Language, document, context.ClientCapabilities, cancellationToken).ConfigureAwait(false);
+        }
 
-            static async Task<Hover> GetHoverAsync(QuickInfoItem info, Document document, ClientCapabilities clientCapabilities, CancellationToken cancellationToken)
+        internal static async Task<Hover?> GetHoverAsync(
+            SemanticModel semanticModel,
+            int position,
+            HostLanguageServices languageServices,
+            CancellationToken cancellationToken)
+        {
+            var quickInfoService = (QuickInfoServiceWithProviders)languageServices.GetRequiredService<QuickInfoService>();
```

Comment:
can you doc taht this is safe because... reasons?  maybe that this is only for c#/vb so we can trust in this?